core/src/main/java/com/amazon/opendistroforelasticsearch/sql/expression/datetime/IntervalClause.java

```diff
@@ -0,0 +1,172 @@
+import org.apache.commons.lang3.tuple.Pair;
+
+@UtilityClass
+public class IntervalClause {
+
+  private static final String MICRO_SECOND = "microsecond";
+  private static final String SECOND = "second";
+  private static final String MINUTE = "minute";
+  private static final String HOUR = "hour";
+  private static final String DAY = "day";
+  private static final String WEEK = "week";
+  private static final String MONTH = "month";
+  private static final String QUARTER = "quarter";
+  private static final String YEAR = "year";
+
+  public void register(BuiltinFunctionRepository repository) {
+    repository.register(interval());
+  }
+
+  private FunctionResolver interval() {
+    return define(BuiltinFunctionName.INTERVAL.getName(),
+        intervalImp(IntervalClause::interval, INTEGER),
+        intervalImp(IntervalClause::interval, LONG),
+        intervalImp(IntervalClause::interval, FLOAT),
+        intervalImp(IntervalClause::interval, DOUBLE));
+  }
+
+  private ExprValue interval(ExprValue value, ExprValue unit) {
+    switch (getStringValue(unit).toLowerCase()) {
+      case MICRO_SECOND:
+        return microsecond(value);
+      case SECOND:
+        return second(value);
+      case MINUTE:
+        return minute(value);
+      case HOUR:
+        return hour(value);
+      case DAY:
+        return day(value);
+      case WEEK:
+        return week(value);
+      case MONTH:
+        return month(value);
+      case QUARTER:
+        return quarter(value);
+      case YEAR:
+        return year(value);
+      default:
+        throw new ExpressionEvaluationException(
+            String.format("interval unit %s is not supported", getStringValue(unit)));
+    }
+  }
+
+  private ExprValue microsecond(ExprValue value) {
+    return new ExprIntervalValue(Duration.ofNanos(getLongValue(value) * 1000));
+  }
+
+  private ExprValue second(ExprValue value) {
+    return new ExprIntervalValue(Duration.ofSeconds(getLongValue(value)));
+  }
+
+  private ExprValue minute(ExprValue value) {
+    return new ExprIntervalValue(Duration.ofMinutes(getLongValue(value)));
+  }
+
+  private ExprValue hour(ExprValue value) {
+    return new ExprIntervalValue(Duration.ofHours(getLongValue(value)));
+  }
+
+  private ExprValue day(ExprValue value) {
+    return new ExprIntervalValue(Duration.ofDays(getIntegerValue(value)));
+  }
+
+  private ExprValue week(ExprValue value) {
+    return new ExprIntervalValue(Period.ofWeeks(getIntegerValue(value)));
+  }
+
+  private ExprValue month(ExprValue value) {
+    return new ExprIntervalValue(Period.ofMonths(getIntegerValue(value)));
+  }
+
+  private ExprValue quarter(ExprValue value) {
+    return new ExprIntervalValue(Period.ofMonths(getIntegerValue(value) * 3));
+  }
+
+  private ExprValue year(ExprValue value) {
+    return new ExprIntervalValue(Period.ofYears(getIntegerValue(value)));
+  }
+
+  /**
+   * Interval argument implementation as function.
+   */
+  private static SerializableFunction
```

Comment:
could we reuse the FunctionDSL.impl?

Reply:
`toString()` method needs to be overridden

Reply:
> `toString()` method needs to be overridden

Where is the `toString` method used? I recall it was required earlier because we didn't have original field name in `SELECT`?